src/egraph.rs

```diff
@@ -202,6 +202,58 @@ impl<L: Language> EGraph<L> {
         pruned
     }
 
+    pub fn fold_constants(&mut self) -> usize {
+        let mut folded = 0;
+        let mut to_add = HashMap::default();
+        // keep track of constants in each class to save some scanning
+        let mut constant_nodes = HashMap::default();
+
+        // look for constants in each class
+        for (id, class) in &self.classes {
+            for node in &class.nodes {
+                if let Expr::Constant(c) = node {
+                    constant_nodes.insert(*id, c.clone());
+                }
+            }
+        }
+
+        // evaluate foldable expressions
+        for (id, class) in &self.classes {
+            for node in &class.nodes {
+                if let Expr::Operator(op, cids) = node {
+                    // get children if they are all constant
+                    let children: Option<Vec<_>> = cids.iter().map(|id| {
+                        if let Some(c) = constant_nodes.get(id) {
+                            Some(c.clone())
```

Comment:
You can use [`Option::clone`](https://doc.rust-lang.org/std/option/enum.Option.html#method.cloned) to clone the inside of an Option.

Reply:
To be clear, I think this whole lambda can be replaced with:
```rust
|id| constant_nodes.get(id).cloned()
```